addons/web/static/src/js/views/graph/graph_renderer.js

```diff
@@ -1,1099 +1,1046 @@
-     */
-    _prepareOptions: function (datasetsCount) {
-        const options = {
-            maintainAspectRatio: false,
-            scales: this._getScaleOptions(),
-            legend: this._getLegendOptions(datasetsCount),
-            tooltips: this._getTooltipOptions(),
-            elements: this._getElementOptions(),
-        };
-        if (this._isRedirectionEnabled()) {
-            options.onClick = this._onGraphClicked.bind(this);
+                const labelIndex = dp.labelIndex;
+                datasetsTmp[datasetLabel].data[labelIndex] = dp.value;
+                datasetsTmp[datasetLabel].domain[labelIndex] = dp.domain;
+            }
+            // sort by origin
+            const datasets = sortBy(Object.values(datasetsTmp), "originIndex");
+            return { datasets, labels };
         }
-        return options;
-    },
-    /**
-     * Determine how to relabel a label according to a given origin.
-     * The idea is that the getLabel function is in general not invertible but
-     * it is when restricted to the set of dataPoints coming from a same origin.
 
-     * @private
-     * @param {Array} label
-     * @param {Array} originIndex
-     * @returns {string}
-     */
-    _relabelling: function (label, originIndex) {
-        if (label.isNoData) {
-            return label[0];
-        }
-        var i = this.state.comparisonFieldIndex;
-        if (_.contains(['bar', 'line'], this.state.mode) && i === 0) {
-            // here label is an array of length 1 and contains a number
-            return this.dateClasses.representative(label, originIndex) || '';
-        } else if (this.state.mode === 'pie' && i === 0) {
-            // here label is an array of length at least one containing string or numbers
-            var labelCopy = label.slice(0);
-            if (originIndex !== undefined) {
-                labelCopy.splice(i, 1, this.dateClasses.representative(label[i], originIndex));
-            } else {
-                labelCopy.splice(i, 1, this.dateClasses.dateClassMembers(label[i]));
+        /**
+         * Computes various objects based on the props parameter and returns
+         * the graph configuration object.
+         * @private
+         * @param {Object} props
+         * @returns {Object}
+         */
+        _prepareGraphConfig(props) {
+            // Extract some props to prepare some information before the new
+            // props have been applied.
+            this.comparisonFieldIndex = props.comparisonFieldIndex;
+            this.dataPoints = props.dataPoints;
+            this.disableLinking = props.disableLinking;
+            this.domains = props.domains;
+            this.fields = props.fields;
+            this.measure = props.fields[props.measure];
+            this.mode = props.mode;
+            this.orderBy = props.orderBy;
+            this.origins = props.origins;
+            this.processedGroupBy = props.processedGroupBy;
+            this.stacked = props.stacked;
```

Comment:
Not sure this is a good idea. You'll get in a situation where this.props.stacked (for instance) is X, and this.stacked is Y, which is weird. If they are not applied yet, why don't you simply pass props as argument? That way, you remove a bunch of attributes, the duplication, and sometimes the fact that a data is available from 2 places with a different value, which can confuse people.

Reply:
This is to avoid carrying a bunch of props accross several functions: I think it is simpler to store and update these values on each update rather than overloading the parameters of most graph functions with a `props` param. We can discuss this further if you think this is a bad idea.